pkg/server/delta/v3/server.go

```diff
@@ -158,34 +157,27 @@ func (s *server) processDelta(str stream.DeltaStream, reqCh <-chan *discovery.De
 			}
 
 			typeURL := req.GetTypeUrl()
-			state, ok := watches.deltaStreamStates[typeURL]
-			if !ok {
-				// Initialize the state if we haven't already.
-				state = stream.NewStreamState()
-
-				// Since there was no previous state, we know we're handling the first request of this type on this stream
-				for r, v := range req.InitialResourceVersions {
-					state.ResourceVersions[r] = v
-				}
-				state.Wildcard = len(req.GetResourceNamesSubscribe()) == 0
-
-				// This means that we're handling the first request for this type on the stream
-				watches.deltaStreamStates[typeURL] = state
-			}
-			// versionMap := state.GetResourceVersions()
 
 			// cancel existing watch to (re-)request a newer version
-			watch, ok := watches.deltaResponses[typeURL]
+			watch, ok := watches.deltaWatches[typeURL]
 			if !ok {
-				// We must signal goroutine termination to prevent a race between the cancel closing the watch
-				// and the producer closing the watch.`	`
-				if terminate, exists := watches.deltaTerminations[typeURL]; exists {
-					close(terminate)
-				}
-				if watch.cancel != nil {
-					watch.cancel()
+				// Check state for the stream, if we don't have one then initialize
+				state := watches.deltaWatches[typeURL].state
+				if state.ResourceVersions == nil {
```

Comment:
Well we watch can have an uninitialized resource version map as the stream state if this is the first request coming through. I added that to avoid any nil assignments and I also use it as a signal to parse through the InitialResourceVersions if we receive any.

Reply:
But we're making two reads of the same thing: both at L159 and at L162, so we should get the same result both times right? And if !ok then we'd get back a default constructed state, which would always have `state.ResourceVersions == nil`? 

Also, don't we need a line where we write this new state object back into `watches.deltaWatches`?

Reply:
Oh shoot I'm sorry I see what you're saying now. Yes you're right that nil check would be mute since it's always going to be nil in its default state.

The change I'll push actually gets rid of the state object so we won't need to write back. I'll just do all the modifications to `watch.state`